### PR TITLE
Berry Expansion

### PR DESCRIPTION
**Pull request by Bassoonian**
Introducing the Berry Expansion, which aims to allow for flexibility in Berry dynamics. Toggling option doesn't break saves (but may cause minimal Berry output for existing plants, obviously).

This is a draft until all features are finished, but intermediate reviewing would be highly appreciated.

## Features
- [X] Berry Mutations (from XY): planting a Berry tree next to another has a `BERRY_MUTATION_CHANCE` (25% by default) chance of causing a mutation (as dictated by `sBerryMutations`). Mutations mean that besides the usual output, the Berry plant will have a single Berry of the mutation (e.g. planting a Iapapa Berry next to a Mago Berry will cause it to have a single Pomeg Berry on top of its usual output).
- [x] Easier Berry Debugging: add berry manipulation functions to the debug menu to allow for forced growth and more
- [x] Mulch (from Gen IV and XY): using it on soil affects the growth, watering and production values of the plants
- [x] Gradient watering (from Gen IV and XY): rather than keeping track of if each stage has been watered like Gen III, the humidity of the soil is kept track of separately.
- [x] Weeding (from XY): Berries may require unweeding for additional produce.
- [x] Pests (from XY): Bug-type Pokémon may appear to feast on your plants.
- [X] Customisable stages: XY has six stages rather than four, so with this easy toggle you can choose the amount of stages without influencing the growth time.
- [x] Growth configs: pick a generation whose Berry growth rates to use
- [x] Yield configs: pick a generation whose Berry yields to use

## Images
Berry Mutations: (uses Cheri rather than Pomeg as part of internal testing, works as intended when tested now)
berrymutations

Debug Menu - clear planted Berries:
berrydebugclearplants

Debug Menu - force planted Berry growth:
berrydebugforceplants

## Issue(s) that this PR introduces
- The gradient moisture is not reflected in the actual berry soil (palette shenanigans and out of my comfort zone, sorry!)
- Weeds are not reflected graphically (see above)
- Pests are not reflected graphically. This could be as easy as a sped up wiggling animation, but it's something I personally don't want to bother with, sorry :)

## **Discord contact info**
bassoonian

### CODE REVIEW

**Comment by Bassoonian** [comment]
This should now be ready for review! It got very complicated very quickly with all the non-vanilla cross-generational interactions and having to preserve save compatibility, but all of the features should be functional. Please do your best to break it and also please roast my awful code, thanks :)

File: include/global.berry.h [hunk 1]
```diff
@@ -62,16 +75,21 @@ struct BattleEnigmaBerry
...
+    u8 berryYield:5;
+    u8 pests:1;
+    u8 mutationB:2;
     u8 regrowthCount:4;
-    u8 watered1:1;
-    u8 watered2:1;
-    u8 watered3:1;
-    u8 watered4:1;
+    u8 watered:4; // Used to keep track of bonuses in case of gradient watering
+    u16 moistureLevel:7;
```
Comment (GraionDilach) [inline include/global.berry.h:90]: I haven't started testing this yet, but why doesn't this added word break saves? BerryTree is stored in SaveBlock1 afterall.

File: src/berry.c [hunk 2]
```diff
@@ -1849,3 +2403,210 @@ void SetBerryTreesSeen(void)
...
+}
+
+static u16 GetBerryPestSpecies(u8 berryId)
+{
+#if OW_BERRY_PESTS == TRUE
+    const struct Berry *berry = GetBerryInfo(berryId);
+    switch(berry->color)
+    {
+        case BERRY_COLOR_RED:
+            return SPECIES_LEDYBA;
```
Comment (GraionDilach) [inline src/berry.c:2546]: Shouldn't this block check if the families are enabled?

Comment (Bassoonian) [reply src/berry.c:2546]: Correct, this PR was made before species refactor was pushed. I'll add the checks shortly

File: include/global.berry.h (hunk 1 above)
Comment (Bassoonian) [reply include/global.berry.h:90]: Because the struct is padded by the compiler and this added word just recycles said padding.

**Comment by Bassoonian** [comment]
Updated to the latest state of upcoming, still awaiting review!

**Comment by Ninjdai1** [comment]
I'm unsure of what happened, but:
berries

I'll try to reproduce that

**Comment by Bassoonian** [comment]
> I'm unsure of what happened, but (gif here)

Issue resolved!

**Comment by Ninjdai1** [comment]
Should growing berry trees with the debug menu as they're fully grown reset them back to just planted stage ?
berries_4

**Comment by Bassoonian** [comment]
> Should growing berry trees with the debug menu as they're fully grown actually reset them back to just planted stage ? berries_4 [ berries_4 ](https://private-user-images.githubusercontent.com/65647523/292882511-4b7174ca-994a-462a-8901-64b10053943c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2ODIyMDUsIm5iZiI6MTcwMzY4MTkwNSwicGF0aCI6Ii82NTY0NzUyMy8yOTI4ODI1MTEtNGI3MTc0Y2EtOTk0YS00NjJhLTg5MDEtNjRiMTAwNTM5NDNjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI3VDEyNTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlMzc0MGZhODVkMjcxZGJmYjBlMzljYTA1MTYwZjYzY2M3ZWMzNzRjZDQ0NjZkOTZhN2NiZmJhMzM3Y2RiYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RUjugbYiKdVWK9L5XSBA0y_sDtjN7lRX0FNI4lL5Y84) [ ](https://private-user-images.githubusercontent.com/65647523/292882511-4b7174ca-994a-462a-8901-64b10053943c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2ODIyMDUsIm5iZiI6MTcwMzY4MTkwNSwicGF0aCI6Ii82NTY0NzUyMy8yOTI4ODI1MTEtNGI3MTc0Y2EtOTk0YS00NjJhLTg5MDEtNjRiMTAwNTM5NDNjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI3VDEyNTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlMzc0MGZhODVkMjcxZGJmYjBlMzljYTA1MTYwZjYzY2M3ZWMzNzRjZDQ0NjZkOTZhN2NiZmJhMzM3Y2RiYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RUjugbYiKdVWK9L5XSBA0y_sDtjN7lRX0FNI4lL5Y84)

It's partially intentional, as whenever the timer until the next stage for grown berries reaches zero, they die and (if not yet done too often) reseed themselves (which is vanilla behaviour). The function I call to force growth includes that functionality, so I just left it as is.

**Comment by DizzyEggg** [comment]
Conflicts 👀

**Comment by Bassoonian** [comment]
Conflicts resolved! Also replaced berry descriptions with compound strings, removed an unused table and removed an unused function.

**Review by DizzyEggg — APPROVED** [review]

**Pull request merged**